tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,607 @@
+package tensorflow
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+
+	pb "github.com/tensorflow/tensorflow/tensorflow/contrib/go/proto"
+)
+
+const (
+	cOpsProtobufDefsPath = "/usr/local/tensorlow/ops.pbtxt"
+)
+
+// Graph Representation of the computation graph.
+type Graph struct {
+	def *pb.GraphDef
+
+	availableOps map[string]*pb.OpDef
+	constants    map[string]*Tensor
+	variables    map[string]*Tensor
+}
+
+// GraphNode Representation of one of the nodes of the TensorFlow Graph a
+// node takes zero or more Tensors, performs some computation, and
+// produces zero or more Tensors.
+type GraphNode struct {
+	ref          *pb.NodeDef
+	def          *pb.NodeDef
+	outDataTypes map[string]DataType
+}
+
+// NewGraph Returns an initialized instance of the Graph struct.
+func NewGraph() *Graph {
+	return &Graph{
+		def:          new(pb.GraphDef),
+		availableOps: make(map[string]*pb.OpDef),
+		constants:    make(map[string]*Tensor),
+		variables:    make(map[string]*Tensor),
+	}
+}
+
+// NewGraphFromText Returns a new graph populated with the deserialization of
+// the provided graph string.
+func NewGraphFromText(graphStr string) (gr *Graph, err error) {
+	gr = NewGraph()
+	err = proto.UnmarshalText(graphStr, gr.def)
+
+	return
+}
+
+// LoadGraphFromFile Loads a Graph from the file on the specified path.
+func LoadGraphFromFile(path string) (gr *Graph, err error) {
+	graphStr, err := ioutil.ReadFile(path)
+	if err != nil {
+		return
+	}
+
+	gr = NewGraph()
+	err = proto.Unmarshal(graphStr, gr.def)
+
+	return
+}
+
+// LoadGraphFromTextFile Loads a Graph as plain text from the file on the specified
+// path.
+func LoadGraphFromTextFile(path string) (gr *Graph, err error) {
+	graphStr, err := ioutil.ReadFile(path)
+	if err != nil {
+		return
+	}
+
+	return NewGraphFromText(string(graphStr))
+}
+
+// Op Adds a new Node to the Graph with the specified operation, this
+// operation performs some internal check of the specified and expected
+// attributes for the operation and try to deduct the corresponding DataTypes
+// in case of they are not specified.
+func (gr *Graph) Op(opName string, name string, input []*GraphNode, device string, attrs map[string]interface{}) (node *GraphNode, err error) {
+	if err = gr.loadAvailableOps(); err != nil {
+		return
+	}
+
+	if op, ok := gr.availableOps[strings.ToLower(opName)]; ok {
```

Comment:
I think it would really clarify this code to reverse the test here.  If the op is *not* in the list of available ops, immediately return the ErrOperationNotFound error.  Otherwise, proceed into the function as before.  (This saves you a lot of unnecessary indentation and makes the error vs normal case a little more clear.)